### PR TITLE
Publish recall as a kpi metric

### PR DESCRIPTION
**Pull request by finnroblin**
### Description
Displays recall as part of the final benchmarking stats.
i.e.: 
```
------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------
            
|                                                         Metric |                     Task |       Value |   Unit |
|---------------------------------------------------------------:|-------------------------:|------------:|-------:|
|                     Cumulative indexing time of primary shards |                          |     16.1933 |    min |
.............................................................................................      .....................................................................
| Mean recall@k |             prod-queries |        0.99 |        |

|                                                  Mean recall@1 |             prod-queries |           1 |        |
|                           50th percentile recall_k percentiles |             prod-queries |           1 |        |
|                           90th percentile recall_k percentiles |             prod-queries |           1 |        |
|                           99th percentile recall_k percentiles |             prod-queries |           1 |        |
|                          100th percentile recall_k percentiles |             prod-queries |           1 |        |
```

### Issues Resolved
(From workloads repo): [282](https://github.com/opensearch-project/opensearch-benchmark-workloads/issues/282)
While this PR is not a general solution to #199, it does address the specific need mentioned in that issue of displaying the recall metadata as a summary. 

I have added a `kpi_metrics` field to the `GlobalStats` metric collector where additional metrics could be added. Users would then need to add display logic to the `results_publisher.py` file in order for the metrics to show up in the summary. 

### Testing
- [X] New functionality includes testing

I added some unit tests to assert that recall metrics were collected. I also altered the `OsResultsStoreTest` to account for the new `kpi_metrics` field.

I manually tested the `train-test` workload, which contains a vector search operation, and the `no-train-test-index-only` workload, which does not contain a vector search operation. Both workloads ran to completion and displayed the expected statistics (recall for `train-test`, and no recall metrics for `no-train-test-index-only`). 

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: osbenchmark/metrics.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 # not use this file except in compliance with the License.
 # You may obtain a copy of the License at
 #
-#	http://www.apache.org/licenses/LICENSE-2.0
+# 	http://www.apache.org/licenses/LICENSE-2.0
```
Comment (finnroblin) [inline osbenchmark/metrics.py:16]: Not sure why my code formatter keeps on changing code outside of the selected lines. I'll revert this and the other spacing changes to out-of-scope code in the next revision.

File: osbenchmark/metrics.py [hunk 2]
```diff
@@ -2094,6 +2115,18 @@ def add_op_metrics(self, task, operation, throughput, latency, service_time, cli
             doc["meta"] = meta
         self.op_metrics.append(doc)
 
+    def add_kpi_metrics(self, task, operation, recall_at_k_stats, recall_at_1_stats, recall_time_ms_stats, error_rate, duration):
+        self.kpi_metrics.append({
+            "task": task,
+            "operation": operation,
+            "recall@k": recall_at_k_stats,
+            "recall@1":recall_at_1_stats,
+            "recall_time_ms": recall_time_ms_stats,
```
Comment (finnroblin) [inline osbenchmark/metrics.py:2124]: I spoke with @VijayanB and `recall_time_ms` is not a stat we want to display. I'll remove it from the `kpi_metrics` method in the next revision.

File: osbenchmark/worker_coordinator/worker_coordinator.py [hunk 3]
```diff
@@ -847,6 +846,50 @@ def __call__(self, raw_samples):
...
+                            sample.task.name,
+                            sample.operation_name,
+                            sample.operation_type,
+                            sample.sample_type,
+                        )
+                        self.metrics_store.put_value_cluster_level(
+                            name=recall_metric_name,
+                            value=sample.request_meta_data[recall_metric_name],
+                            unit="",
+                            task=sample.task.name,  # todo change unit to segment count unit...
```
Comment (finnroblin) [inline osbenchmark/worker_coordinator/worker_coordinator.py:884]: Will remove todo comment in next revision

File: osbenchmark/metrics.py [hunk 4]
```diff
@@ -1966,6 +1986,7 @@ def single_latency(self, task, operation_type, metric_name="latency"):
 class GlobalStats:
     def __init__(self, d=None):
         self.op_metrics = self.v(d, "op_metrics", default=[])
+        self.kpi_metrics = self.v(d, "kpi_metrics", default=[])
```
Comment (jmazanec15) [inline osbenchmark/metrics.py:1989]: What is the scope for kpi metrics? Aside from recall and maybe ndcg, what else would go in here?

File: osbenchmark/metrics.py (hunk 2 above)
Comment (jmazanec15) [reply osbenchmark/metrics.py:2124]: Can we remove from this PR?

File: osbenchmark/metrics.py (hunk 4 above)
Comment (finnroblin) [reply osbenchmark/metrics.py:1989]: I think any performance metric for k-NN is in scope for this. Right now it seems like recall is the most major one but if k-NN adds ndcg we could also track that in OSB. (The process for adding new metrics would be to calculate/add it to the metadata in the search query runner and then pull it from the metadata into this `kpi_metrics` list). Likewise if we or the anomaly detection plugin want to track false positives rate/specificity.

I mainly wanted to explicitly distinguish between algorithm metrics like recall and `op_metrics` like latency.

File: osbenchmark/metrics.py (hunk 2 above)
Comment (finnroblin) [reply osbenchmark/metrics.py:2124]: Sorry not sure what "this" is referring to. Do you mean remove the `add_kpi_metrics` function (and instead append directly to the `kpi_metrics` list), remove the `recall_time_ms` metric, or remove `kpi_metrics` all together and go a different direction for passing recall from metadata to the metrics store?

File: tests/metrics_test.py [hunk 5]
```diff
@@ -1436,97 +1436,105 @@ def test_store_results(self):
                 "benchmark-version": "0.4.4",
                 "benchmark-revision": "123abc",
                 "environment": "unittest",
-                "test-execution-id": OsResultsStoreTests.TEST_EXECUTION_ID,
+                "test-execution-id": "6ebc6e53-ee20-4b0c-99b4-09697987e9f4",
```
[the remark this answers is not in the document]
Comment (finnroblin) [reply tests/metrics_test.py:1439]: I think the ordering of the json got scrambled when I added the kpi_metrics variable. I agree though that I shouldn't have changed so much; I'll revert most of the unit test and just add a single extra element in the json array with the kpi_metrics and associated data.

File: osbenchmark/metrics.py (hunk 2 above)
Comment (jmazanec15) [reply osbenchmark/metrics.py:2124]: sorry meant the recall_time_ms - not sure if this is needed

File: osbenchmark/metrics.py (hunk 4 above)
Comment (jmazanec15) [reply osbenchmark/metrics.py:1989]: Oh I see. KPI though might include latency right? It seems like a good category here would be `correctness_metrics` or `relevance_metrics`

File: osbenchmark/metrics.py [hunk 6]
```diff
@@ -13,7 +13,7 @@
 # not use this file except in compliance with the License.
 # You may obtain a copy of the License at
 #
-#	http://www.apache.org/licenses/LICENSE-2.0
+#	http://www.apache.org/licenses/LICENSE-2.
```
Comment (VijayanB) [inline osbenchmark/metrics.py:16]: can you remove unrelated formatting changes from this PR?

File: osbenchmark/results_publisher.py [hunk 7]
```diff
@@ -156,6 +156,16 @@ def publish(self):
             metrics_table.extend(self._publish_error_rate(record, task))
             self.add_warnings(warnings, record, task)
 
+        for record in stats.correctness_metrics:
+            task = record["task"]
+
+            keys = record.keys()
+            recall_keys_in_task_dict = "recall@1" in keys and "recall@k" in keys
+            if recall_keys_in_task_dict and record["recall@1"] and record["recall@k"]:
```
Comment (VijayanB) [inline osbenchmark/results_publisher.py:164]: don't check for record['recall@k"], of recall is 0, it will be skipped. just check only for keys

File: osbenchmark/results_publisher.py [hunk 8]
```diff
@@ -200,14 +210,26 @@ def _publish_service_time(self, values, task):
...
+        recall_k = values["recall@k"]
+        recall_1 = values["recall@1"]
+
+        try:
+            return self._join(
+                self._line("Mean recall@k", task, recall_k["mean"], "", lambda v: "%.2f" % v),
+                self._line("Mean recall@1", task, recall_1["mean"], "", lambda v: "%.2f" % v)
+            )
+        except KeyError:
+            return None
```
Comment (VijayanB) [inline osbenchmark/results_publisher.py:223]: why do we need to catch this? can you return from like 214 if keys doesn't exist?

File: osbenchmark/worker_coordinator/worker_coordinator.py [hunk 9]
```diff
@@ -1369,7 +1413,7 @@ def calculate_task_throughput(self, task, current_samples, bucket_interval_secs)
             self.task_stats[task] = ThroughputCalculator.TaskStats(bucket_interval=bucket_interval_secs,
                                                                    sample_type=first_sample.sample_type,
                                                                    start_time=first_sample.absolute_time - first_sample.time_period)
-        current = self.task_stats[task]
+        current = self.task_stats[task] # TaskStats object
```
Comment (VijayanB) [inline osbenchmark/worker_coordinator/worker_coordinator.py:1416]: why do we need this comment here and as part of this PR?

File: osbenchmark/worker_coordinator/worker_coordinator.py [hunk 10]
```diff
@@ -1190,6 +1233,7 @@ def __init__(self, start_timestamp, buffer_size=16384):
     def add(self, task, client_id, sample_type, meta_data, absolute_time, request_start, latency, service_time,
             client_processing_time, processing_time, throughput, ops, ops_unit, time_period, percent_completed,
             dependent_timing=None):
+        self.logger.debug("Logging with metadata: [%s]", meta_data)
```
Comment (VijayanB) [inline osbenchmark/worker_coordinator/worker_coordinator.py:1236]: do we need this?

File: osbenchmark/worker_coordinator/worker_coordinator.py [hunk 11]
```diff
@@ -847,6 +846,50 @@ def __call__(self, raw_samples):
...
+                    if recall_metric_name in sample.request_meta_data:
+                        meta_data = self.merge(
+                            self.workload_meta_data,
+                            self.test_procedure_meta_data,
+                            sample.operation_meta_data,
+                            sample.task.meta_data,
+                            sample.request_meta_data,
+                        )
+
+                        self.logger.debug(
```
Comment (VijayanB) [inline osbenchmark/worker_coordinator/worker_coordinator.py:873]: do we need this to log?

**Review by VijayanB — APPROVED** [review]
LGTM

**Comment by VijayanB** [comment]
@jmazanec15 Can you take a look at this PR? Thanks

File: osbenchmark/worker_coordinator/worker_coordinator.py (hunk 11 above)
Comment (finnroblin) [reply osbenchmark/worker_coordinator/worker_coordinator.py:873]: Thanks for catching this -- It's debug level so it shouldn't show up with normal logging settings. I was using it to figure out how samples worked, so this log has served its purpose and I'll remove it now. In general it doesn't seem like there are any other debug level log statements within OSB, so I will avoid committing them in the future.

**Review by jmazanec15 — APPROVED** [review]
This is awesome! great job @finnroblin !
